### DISCUSSION TITLE
baresip registration with 1und1.de

### DISCUSSION THREAD

**Discussion by chri2** [baresip/baresip#3149] [Q&A]
I'm trying to register to 1und1.de using baresip and getting `403 Contact User und Anrufernummer verschieden (UI Kamailio)` as an error.

My registration looks like this:
```
REGISTER sip:1und1.de SIP/2.0
Via: SIP/2.0/UDP 192.168.180.2:54867;branch=jlk34kjhk4h3k4hk1234h4h;rport
Contact: <sip:49xxxxxxxxxx-0x559e4bd590@192.168.1.1:54867>;expires=3786;+sip.instance="<urn:uuid:aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa>";reg-id=1
Max-Forwards: 70
Route: <sip:49xxxxxxxxxx@1und1.de;lr>
To: <sip:49xxxxxxxxxx@1und1.de>
From: "1&1" <sip:49xxxxxxxxxx@1und1.de>;tag=aaaaaaaaaaaaaaa
Call-ID: aaaaaaaaaaaaaaaa
CSeq: 33680 REGISTER
User-Agent: baresip v1.0.0 (/linux)
Supported: gruu, outbound, path
Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,REFER
Content-Length: 0
```

I found some information about what could cause the error [here](https://markus.heberling.net/2020/04/21/using-1und1-sip-with-freeswitch/) and [here (in German)](https://www.ip-phone-forum.de/threads/grandstream-ucm6102-pbx-1und1-voip-registrierung.299074/post-2379193) and wanted to try to change the `Contact:` header to suite the need of 1und1.de .

* Is there a way to influence the content of the `Contact:` header?
* Does anybody actually have a working config for 1und1.de?
* Any other hint where to look?

**Comment by cspiel1** [comment]
What exactly is the problem for 1und1.de? Is it
- the suffix `0x559e4bd590` of the contact user? (This is useful in order to choose the right UA for incoming SIP requests, e.g. for incoming calls.)
- Or is it the `192.168.1.1:54867` which is different to the host part `1und1.de`? (I think the later would be wrong.)
- Or both? So does 1und1 expect this Contact header: `<sip:49xxxxxxxxxx@1und1.de>;...`

**Reply by chri2** [reply]
I'm sorry, but I do not know more than I wrote. I understood that there are two conditions:
* `Contact:` is the same number as in `From:` - so yes, maybe the suffix is a problem
* `Contact:`  should contain an official IP

**Comment by fosple** [comment]
I had the same issue and figured out the solution:

https://github.com/baresip/baresip/blob/fd02485b82aaa266e386b1b3d569a5f51a44b866/src/ua.c#L1211 causes the issue, with adds a suffix to the user (in the example above the "-0x559e4bd590").

If I change that line to
```
err = re_sdprintf(&ua->cuser, "%r", &ua->acc->luri.user);
```
everything works as expected. 

I know baresip is complicant to the RFC spec. However, if big providers like 1&1 having a different implementation I would argue it would make sense to at least offer a configuration option to support these (by disabling the suffix). Maybe something to look into @cspiel1 ?

Related: 
https://github.com/baresip/baresip/issues/170 as 1und1 also uses Kamailio 
https://github.com/baresip/baresip/issues/1644 - but different provider

**Reply by cspiel1** [reply]
Yes, I agree. We have a patch for this and thought that it is to specific for mainline. But now it maybe is worth to make a PR for discussion, with the maintainers.

**Reply by cspiel1** [reply]
PR: https://github.com/baresip/baresip/pull/3462

**Reply by fosple** [reply]
Wow, that was super fast :) Thanks a lot @cspiel1
